src/node_crypto.cc

```diff
@@ -3450,19 +3449,95 @@ void KeyObject::Export(const FunctionCallbackInfo<Value>& args) {
+MaybeLocal<Value> KeyObjectHandle::ExportPrivateKey(
     const PrivateKeyEncodingConfig& config) const {
-  return WritePrivateKey(env(), asymmetric_key_.get(), config);
+  return WritePrivateKey(env(), data_->GetAsymmetricKey().get(), config);
+}
+
+void NativeKeyObject::New(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK_EQ(args.Length(), 1);
+  CHECK(args[0]->IsObject());
+  KeyObjectHandle* handle = Unwrap<KeyObjectHandle>(args[0].As<Object>());
+  new NativeKeyObject(env, args.This(), handle->Data());
+}
+
+BaseObjectPtr<BaseObject> NativeKeyObject::KeyObjectTransferData::Deserialize(
+        Environment* env,
+        Local<Context> context,
+        std::unique_ptr<worker::TransferData> self) {
+  if (context != env->context()) {
+    THROW_ERR_MESSAGE_TARGET_CONTEXT_UNAVAILABLE(env);
+    return {};
+  }
+
+  Local<Value> handle = KeyObjectHandle::Create(env, data_).ToLocalChecked();
+  Local<Function> key_ctor;
+  switch (data_->GetKeyType()) {
+    case kKeyTypeSecret:
+      key_ctor = env->crypto_key_object_secret_constructor();
+      break;
+    case kKeyTypePublic:
+      key_ctor = env->crypto_key_object_public_constructor();
+      break;
+    case kKeyTypePrivate:
+      key_ctor = env->crypto_key_object_private_constructor();
+      break;
+    default:
+      CHECK(false);
+  }
+
+  Local<Value> key =
+      key_ctor->NewInstance(context, 1, &handle).ToLocalChecked();
+  return BaseObjectPtr<BaseObject>(Unwrap<KeyObjectHandle>(key.As<Object>()));
+}
+
+BaseObject::TransferMode NativeKeyObject::GetTransferMode() const {
+  return BaseObject::TransferMode::kCloneable;
+}
+
+std::unique_ptr<worker::TransferData> NativeKeyObject::CloneForMessaging()
+    const {
+  return std::make_unique<KeyObjectTransferData>(handle_data_);
+}
+
+static void CreateNativeKeyObject(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK_EQ(args.Length(), 1);
+  Local<Value> callback = args[0];
+  CHECK(callback->IsFunction());
+
+  Local<FunctionTemplate> t = env->NewFunctionTemplate(NativeKeyObject::New);
+  t->InstanceTemplate()->SetInternalFieldCount(
+      KeyObjectHandle::kInternalFieldCount);
+  t->Inherit(BaseObject::GetConstructorTemplate(env));
+
+  Local<Value> ctor = t->GetFunction(env->context()).ToLocalChecked();
+
+  Local<Value> recv = Undefined(env->isolate());
+  Local<Array> ret =
+      callback.As<Function>()->Call(env->context(), recv, 1, &ctor)
+      .ToLocalChecked()
+      .As<Array>();
+  env->set_crypto_key_object_secret_constructor(
+      ret->Get(env->context(), 1).ToLocalChecked().As<Function>());
+  env->set_crypto_key_object_public_constructor(
+      ret->Get(env->context(), 2).ToLocalChecked().As<Function>());
+  env->set_crypto_key_object_private_constructor(
+      ret->Get(env->context(), 3).ToLocalChecked().As<Function>());
+  args.GetReturnValue().Set(ret);
```

Comment:
```suggestion
  Local<Value> ret_v;
  if (callback.As<Function>()->Call(
          env->context(), recv, 1, &ctor).ToLocal(&ret)) {
    return;
  }
  Local<Array> ret = ret_v.As<Array>();
  Local<Value> ctor;
  if (!ret->Get(env->context(), 1).ToLocal(&ctor)) return;
  env->set_crypto_key_object_secret_constructor(ctor.As<Function>());
  if (!ret->Get(env->context(), 2).ToLocal(&ctor)) return;
  env->set_crypto_key_object_public_constructor(ctor.As<Function>());
  if (!ret->Get(env->context(), 3).ToLocal(&ctor)) return;
  env->set_crypto_key_object_private_constructor(ctor.As<Function>());
  args.GetReturnValue().Set(ret);
```

(In particular the `->Call()` result can definitely be empty here)

Reply:
This function is only called once by `lib/internal/crypto/keys.js`. In my opinion, any error would be a bug in Node.js and should terminate the process. Happy to change it if you disagree.

Reply:
@tniessen The `->Call()` call can always fail because another thread might terminate execution, either through `Stop()` or `worker.terminate()`. For the other ones, I don’t *think* it’s an issue, but I do feel that we should always handle exceptions properly if possible instead of just crashing, if only because that sets a good precedent for other code.